{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b55cf2d4-c22f-4e65-8357-936b7df7aecb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from math import ceil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f4c63766-f062-4ae2-99fa-20f27e1e8ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# f to fetch data from bam\n",
    "def genes_extractor(samfile,\n",
    "                    genes_bam,\n",
    "                    fetch_df,\n",
    "                    index_file = True,\n",
    "                    border_shift = 100000):\n",
    "    if not index_file:\n",
    "        pysam.index(samfile)\n",
    "    \n",
    "    samfile = pysam.AlignmentFile(samfile, 'rb')\n",
    "    genes_bam = pysam.AlignmentFile(genes_bam, 'wb', template=samfile)\n",
    "    \n",
    "    d = []\n",
    "    \n",
    "    for index, gene in fetch_df.iterrows():\n",
    "        start = gene['start'] - border_shift\n",
    "        stop = gene['stop'] + border_shift\n",
    "        if start < 0:\n",
    "            start = 0\n",
    "        for read in samfile.fetch(gene['chr'], start, stop):\n",
    "            genes_bam.write(read)\n",
    "        for pileupcolumn in samfile.pileup(gene['chr'], start, stop):\n",
    "            d.append(\n",
    "                    {\n",
    "                        'position': pileupcolumn.pos,\n",
    "                        'read_count': pileupcolumn.n,\n",
    "                        'chromosome': gene['chr'],\n",
    "                        'gene': gene['gene']\n",
    "                    }\n",
    "                )\n",
    "            \n",
    "    df_out = pd.DataFrame(d)\n",
    "    df_out.drop_duplicates(subset=['position'], inplace = True)\n",
    "    \n",
    "    genes_bam.close()\n",
    "    samfile.close()\n",
    "    \n",
    "    return df_out\n",
    "\n",
    "# f for bins creation (of a specidic length)\n",
    "def binner(seq_length,\n",
    "           along,\n",
    "           bin_size = 5000):\n",
    "    bins = np.concatenate([([i]*bin_size) for i in along], axis=0)[:seq_length]\n",
    "    return bins\n",
    "\n",
    "def round_25(x):\n",
    "    return round(x*4)/4\n",
    "\n",
    "def main(fetch_df_path,\n",
    "         samfile_normal,\n",
    "         samfile_tumor,\n",
    "         genes_bam_normal,\n",
    "         genes_bam_tumor,\n",
    "         path_to_image,\n",
    "         purity,\n",
    "         ploidy,\n",
    "         path_to_result,\n",
    "         index_file = True,\n",
    "         bin_size = 5000,\n",
    "         rolling_window = 10000):\n",
    "    \n",
    "    \n",
    "    rolling_window = rolling_window\n",
    "    bin_size = bin_size\n",
    "    purity = purity\n",
    "    ploidy = ploidy\n",
    "    \n",
    "    fetch_df = pd.read_csv(fetch_df_path, sep='\\t', header=None, names=['chr', 'start', 'stop', 'gene'])\n",
    "\n",
    "    \n",
    "    df_normal = genes_extractor(samfile_normal, genes_bam=genes_bam_normal, index_file=index_file, fetch_df=fetch_df)\n",
    "    df_tumor = genes_extractor(samfile_tumor, genes_bam=genes_bam_tumor, index_file=index_file, fetch_df=fetch_df)\n",
    "    \n",
    "    # merging tumor and normal data\n",
    "    df_total = df_tumor.merge(df_normal, left_on='position', right_on='position',\n",
    "              suffixes=('_tumor', '_normal'), how='outer')\n",
    "    \n",
    "    # counting depth ratio and median rolling\n",
    "    df_total['depth_ratio'] = np.log2((1-purity) + purity * ((df_total['read_count_tumor'] / df_total['read_count_normal']) + 0.5) / ploidy)\n",
    "    df_total['median_rolling_depth_ratio'] = df_total['depth_ratio'].rolling(rolling_window, min_periods=1, center=True).median()\n",
    "    \n",
    "    # binning\n",
    "    bins = binner(df_total.shape[0], \n",
    "                  along = range(ceil(df_total.shape[0] / bin_size)), \n",
    "                  bin_size = bin_size)\n",
    "    \n",
    "    df_total['bin'] = bins\n",
    "    \n",
    "    # generating median on rolling\n",
    "    df_group = df_total[['position', 'median_rolling_depth_ratio', 'bin']].groupby(['bin'])['median_rolling_depth_ratio'].median()\n",
    "    df_group = df_group.reset_index() \n",
    "    \n",
    "    median_rolling_binning_depth_ratio = binner(df_total.shape[0], \n",
    "                                                 along = df_group['median_rolling_depth_ratio'], \n",
    "                                                 bin_size = bin_size)\n",
    "    \n",
    "    df_total['median_rolling_binning_depth_ratio'] = median_rolling_binning_depth_ratio\n",
    "\n",
    "    # counting CNVs\n",
    "    df_group_cnv = df_total[['position', 'median_rolling_binning_depth_ratio', 'gene_tumor']].groupby(['gene_tumor'])['median_rolling_binning_depth_ratio'].median()\n",
    "    df_group_cnv = df_group_cnv.reset_index()\n",
    "    df_group_cnv['copy_num'] =  2 ** (df_group_cnv['median_rolling_binning_depth_ratio'] + 1)\n",
    "    df_group_cnv_median = df_group_cnv['copy_num'].median()\n",
    "    df_group_cnv['copy_num_corrected'] = round_25(df_group_cnv['copy_num']/df_group_cnv_median)\n",
    "    df_group_cnv.to_csv(f'{path_to_result}/result.tsv', sep='\\t', index=None)\n",
    "    \n",
    "    fig, axs = plt.subplots(len(fetch_df['gene']))\n",
    "    fig.set_size_inches(20, 200)\n",
    "    \n",
    "    for i, gene in enumerate(fetch_df['gene']):\n",
    "        target_gene = gene\n",
    "        # median rolling depth ratio\n",
    "        axs[i].scatter(df_total[df_total['gene_tumor'] == target_gene]['position'], df_total[df_total['gene_tumor'] == target_gene]['median_rolling_depth_ratio'], alpha=0.1, s=0.001, c='grey')\n",
    "        # bins\n",
    "        axs[i].scatter(df_total[df_total['gene_tumor'] == target_gene]['position'], df_total[df_total['gene_tumor'] == target_gene]['median_rolling_binning_depth_ratio'], alpha=0.5, s=0.001, c='violet')\n",
    "    \n",
    "        axs[i].axvline(x = fetch_df[fetch_df['gene'] == target_gene]['start'].iloc[0], color = 'r')\n",
    "        axs[i].axvline(x = fetch_df[fetch_df['gene'] == target_gene]['stop'].iloc[0], color = 'r')\n",
    "        axs[i].set_title(gene)\n",
    "         \n",
    "    plt.savefig(path_to_image)\n",
    "    plt.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "90f51420-eb4f-493e-9d53-95895bf345ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "main(fetch_df_path = 'interesting_genes.bed',\n",
    "    samfile_normal = '../WES-normal.bam',\n",
    "    genes_bam_normal = '../genes.bam',\n",
    "    samfile_tumor = '../WES-tumor.bam',\n",
    "    genes_bam_tumor = '../genes_tumor.bam',\n",
    "    path_to_result = '.',\n",
    "    purity = 0.71,\n",
    "    ploidy = 2.61,\n",
    "    path_to_image = 'genes.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9517c8df-16f5-407c-80bb-fe92796eab90",
   "metadata": {},
   "outputs": [],
   "source": [
    "main(fetch_df_path = 'interesting_genes.bed',\n",
    "    samfile_normal = '../patients/patient5/WES-normal.bam',\n",
    "    genes_bam_normal = '../patients/patient5/genes.bam',\n",
    "    samfile_tumor = '../patients/patient5/WES-tumor.bam',\n",
    "    genes_bam_tumor = '../patients/patient5/genes_tumor.bam',\n",
    "    path_to_image = '../patients/patient5/genes.png',\n",
    "    path_to_result = '../patients/patient5/',\n",
    "    purity = 0.37,\n",
    "    ploidy = 3.70,\n",
    "    index_file = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8aef08ac-c255-4de9-a95f-a4506889337a",
   "metadata": {},
   "outputs": [],
   "source": [
    "main(fetch_df_path = 'interesting_genes.bed',\n",
    "    samfile_normal = '../patients/patient7/WES-normal.bam',\n",
    "    genes_bam_normal = '../patients/patient7/genes.bam',\n",
    "    samfile_tumor = '../patients/patient7/WES-tumor.bam',\n",
    "    genes_bam_tumor = '../patients/patient7/genes_tumor.bam',\n",
    "    path_to_image = '../patients/patient7/genes.png',\n",
    "    path_to_result = '../patients/patient7/',\n",
    "    purity = 0.6,\n",
    "    ploidy = 4,\n",
    "    index_file = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "54a3846a-de0a-424b-89cd-5308e4022ae1",
   "metadata": {},
   "outputs": [],
   "source": [
    "main(fetch_df_path = 'interesting_genes.bed',\n",
    "    samfile_normal = '../patients/patient1/WES-normal.bam',\n",
    "    genes_bam_normal = '../patients/patient1/genes.bam',\n",
    "    samfile_tumor = '../patients/patient1/WES-tumor.bam',\n",
    "    genes_bam_tumor = '../patients/patient1/genes_tumor.bam',\n",
    "    path_to_image = '../patients/patient1/genes.png',\n",
    "    path_to_result = '../patients/patient1/',\n",
    "    purity = 0.37,\n",
    "    ploidy = 3.70,\n",
    "    index_file = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bca4e496-241e-45ca-8f9e-2d68cb5ffd8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e7d2e19-a339-4631-9b72-4d20f70657e2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
